### PR TITLE
Resplit out of place

### PR DESCRIPTION
**Pull request by Cdebus**
Addresses Issue #232 
- In dndarray.py, resplit_(axis) is in-place operation
- In manipulations.py, resplit(tensor, axis) is out-of-place operation
- Test for correct sorting of tensors after splitting and unsplitting
- Changes from PR #342 already included (otherwise resplit doesn't work)

### CODE REVIEW

File: heat/core/communication.py [hunk 1]
```diff
@@ -422,22 +422,28 @@ def __allgather_like(self, func, sendbuf, recvbuf, send_axis, recv_axis, **kwarg
             recv_axis_permutation[0], recv_axis_permutation[recv_axis] = recv_axis, 0
             recvbuf = recvbuf.permute(*recv_axis_permutation)
 
-        # prepare buffer objects
+          # prepare buffer objects
```
Comment (Markus-Goetz) [inline heat/core/communication.py]: remove space

File: heat/core/communication.py [hunk 2]
```diff
@@ -422,22 +422,28 @@ def __allgather_like(self, func, sendbuf, recvbuf, send_axis, recv_axis, **kwarg
...
             mpi_recvbuf = recvbuf
         else:
             mpi_recvbuf = self.as_buffer(recvbuf, recv_counts, recv_displs)
+
             if recv_counts is None:
                 mpi_recvbuf[1] //= self.size
 
-        # perform the scatter operation
+
+         # perform the scatter operation
```
Comment (Markus-Goetz) [inline heat/core/communication.py]: remove leading space

File: heat/core/communication.py [hunk 3]
```diff
@@ -422,22 +422,28 @@ def __allgather_like(self, func, sendbuf, recvbuf, send_axis, recv_axis, **kwarg
...
         if sendbuf is  MPI.IN_PLACE or not isinstance(sendbuf, torch.Tensor):
             mpi_sendbuf = sendbuf
         else:
             mpi_sendbuf = self.as_buffer(sendbuf, send_counts, send_displs)
+
+
             if send_counts is not None:
                 mpi_sendbuf[1] = mpi_sendbuf[1][0][self.rank]
 
+
```
Comment (Markus-Goetz) [inline heat/core/communication.py]: superfluous new lines

File: heat/core/dndarray.py [hunk 4]
```diff
@@ -1902,7 +1902,7 @@ def numpy(self):
         T1.numpy()
         """
        
-        return self.resplit(None)._DNDarray__array.cpu().numpy()
+        return self.resplit_(None)._DNDarray__array.cpu().numpy()
```
Comment (Markus-Goetz) [inline heat/core/dndarray.py]: This should remain resplit() as it is the numpy copy. The original dndarray should not be modified

File: heat/core/logical.py [hunk 5]
```diff
@@ -127,13 +127,13 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
 
     # If only one of the tensors is distributed, unsplit/gather it
     if (x.split is not None) and (y.split is None):
-        x.resplit(axis=None)
+        x.resplit_(axis=None)
```
Comment (Markus-Goetz) [inline heat/core/logical.py]: Should also be an out-of-place operation I believe?

File: heat/core/logical.py [hunk 6]
```diff
@@ -127,13 +127,13 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
...
     if (x.split is None) and (y.split is not None):
-        y.resplit(axis=None)
+        y.resplit_(axis=None)
```
Comment (Markus-Goetz) [inline heat/core/logical.py]: Should also be an out-of-place operation I believe?

File: heat/core/logical.py [hunk 7]
```diff
@@ -127,13 +127,13 @@ def allclose(x, y, rtol=1e-05, atol=1e-08, equal_nan=False):
...
 
     # If both x and y are split, but along different axes, y is redistributed to be split along the same axis as x
     if (x.split is not None) and (y.split is not None) and (x.split != y.split):
-        y.resplit(axis=x.split)
+        y.resplit_(axis=x.split)
```
Comment (Markus-Goetz) [inline heat/core/logical.py]: Should also be an out-of-place operation I believe?

File: heat/core/manipulations.py [hunk 8]
```diff
@@ -12,7 +12,8 @@
     'concatenate',
     'expand_dims',
     'squeeze',
-    'unique'
+    'unique',
+    'resplit'
```
Comment (Markus-Goetz) [inline heat/core/manipulations.py]: alphabetical order

File: heat/core/manipulations.py [hunk 9]
```diff
@@ -645,10 +646,61 @@ def unique(a, sorted=False, return_inverse=False, axis=None):
...
+    nodes as well as the definition of new axis along which the tensor is split without changes to the values.
+
+    WARNING: this operation might involve a significant communication overhead. Use it sparingly and preferably for
+    small tensors.
+
+    Parameters
+    ----------
+    a    : ht.DNDarray
+        The tensor from which to resplit
+
```
Comment (Markus-Goetz) [inline heat/core/manipulations.py]: superfluous newline

File: heat/core/statistics.py [hunk 10]
```diff
@@ -338,19 +338,19 @@ def maximum(x1, x2, out=None, **kwargs):
     # apply split semantics
     if x1.split is not None or x2.split is not None:
         if x1.split == None:
-            x1.resplit(x2.split)
+            x1.resplit_(x2.split)
```
Comment (Markus-Goetz) [inline heat/core/statistics.py]: All these should remain out-of-place operations?

File: heat/core/tests/test_manipulations.py [hunk 11]
```diff
@@ -565,3 +565,100 @@ def test_unique(self):
         data_split_zero = ht.array(torch_array, split=0)
         res, inv = ht.unique(data_split_zero, return_inverse=True, sorted=True)
         self.assertTrue(torch.equal(inv, exp_inv.to(dtype=inv.dtype)))
+
+    def test_resplit(self):
+
```
Comment (Markus-Goetz) [inline heat/core/tests/test_manipulations.py]: superfluous new lines

File: heat/core/dndarray.py (hunk 4 above)
Comment (Cdebus) [reply heat/core/dndarray.py]: whoopsi, find and replace error ^^

**Comment by Markus-Goetz** [comment]
Does not build yet for exp()

**Comment by Cdebus** [comment]
Should be fixed now

**Comment by Markus-Goetz** [comment]
Buggy codecov reporting. A direct codecov run reveals sufficient coverage of the new code

**Pull request merged**